tiledb/ml/readers/_tensor_schema.py

```diff
@@ -411,13 +429,21 @@ def argdiff(a: np.ndarray) -> np.ndarray:
 
 
 class KeyDimQuery:
-    def __init__(self, array: tiledb.Array, key_dim_index: int, **kwargs: Any):
+    def __init__(self, array: tiledb.Array, key_dim_index: int, secondary_slices: Dict[int, Union[int, slice, Sequence[int]]], **kwargs: Any):
         self._multi_index = array.query(**kwargs).multi_index
-        self._leading_dim_slices = (slice(None),) * key_dim_index
+        self._key_dim_index = key_dim_index
+        self._slices = [slice(None),] * array.ndim
+        for secondary_index, secondary_slice in secondary_slices.items():
+            # key_dim_index got swapped with 0th index in dimension indexing convention, so swap back
+            if secondary_index == self._key_dim_index:
+                secondary_index = 0
+            self._slices[secondary_index] = secondary_slice
+
 
     def __getitem__(self, key_dim_slice: slice) -> Dict[str, np.ndarray]:
         """Query the TileDB array by `dim_key=key_dim_slice`."""
+        slices = (*self._slices[:self._key_dim_index], key_dim_slice, *self._slices[self._key_dim_index + 1:])
```

Comment:
The two sublists (`self._slices[:self._key_dim_index]` and `self._slices[self._key_dim_index + 1:]`) can be stored as attributes in `__init__` to avoid redundant slicing on every `__getitem__` call.  So just replace `_slices` with `_leading_selectors` and `_trailing_selectors`.